Python/fileutils.c

```diff
@@ -1804,6 +1812,50 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
             }
         }
     }
+
+    // In non-blocking mode, a write that exceeds the available size of a pipe
+    // fails without writing anything. Limiting writes to the pipe size allows
+    // a buffered write to succeed eventually, as the pipe is read.
+    HANDLE hfile = _Py_get_osfhandle(fd);
+    if (hfile == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+    DWORD mode, pipe_size;
+    if (gil_held) {
+        Py_BEGIN_ALLOW_THREADS
+        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
+            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
+                                     NULL, 0) &&
+            mode & PIPE_NOWAIT)
+        {
+            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
+            // CreatePipe() includes this access for the write handle.
+            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
+                pipe_size = 4096;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+        }
+        Py_END_ALLOW_THREADS
+    }
+    else {
+        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
+            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
+                                     NULL, 0) &&
+            mode & PIPE_NOWAIT)
+        {
+            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
+            // CreatePipe() includes this access for the write handle.
+            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
+                pipe_size = 4096;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+        }
+    }
```

Comment:
I overlooked outbound and duplex pipes, which allow writes to the server end of the pipe. Writing to the server end is limited to the out-buffer size/quota. The `GetNamedPipeInfo()` call needs to get the pipe flags. If `PIPE_SERVER_END` is set, use the out-buffer size, else use the in-buffer size.

```suggestion
    DWORD pipe_mode, pipe_flags, pipe_size, pipe_insize, pipe_outsize;
    if (gil_held) {
        Py_BEGIN_ALLOW_THREADS
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            pipe_mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
                                  &pipe_insize, NULL))
            {
                pipe_size = 4096;
            }
            else {
                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
                                pipe_insize;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
        Py_END_ALLOW_THREADS
    }
    else {
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            pipe_mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
                                  &pipe_insize, NULL))
            {
                pipe_size = 4096;
            }
            else {
                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
                                pipe_insize;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
    }
```